src/server/tran_server.cpp

```diff
@@ -609,6 +615,74 @@ tran_server::connection_handler::is_connected ()
   return m_state == state::CONNECTED;
 }
 
+bool
+tran_server::connection_handler::is_idle ()
+{
+  auto slock = std::shared_lock<std::shared_mutex> { m_state_mtx };
+  return m_state == state::IDLE;
+}
+
+tran_server::ps_connector::ps_connector (tran_server &ts)
+  : m_ts { ts }
+  , m_daemon { nullptr }
+  , m_terminate { false }
+{
+}
+
+tran_server::ps_connector::~ps_connector ()
+{
+  terminate ();
+}
+
+void
+tran_server::ps_connector::start ()
+{
+  assert (m_terminate.load () == false);
+  /* After init_page_server_hosts() */
+  assert (m_ts.m_page_server_conn_vec.empty () == false);
+
+  auto func_exec = std::bind (&tran_server::ps_connector::try_connect_to_all_ps, std::ref (*this), std::placeholders::_1);
+  auto entry = new cubthread::entry_callable_task (std::move (func_exec));
+
+  constexpr std::chrono::seconds five_sec { 5 };
+  cubthread::looper loop (five_sec);
+  m_daemon = cubthread::get_manager ()->create_daemon (loop, entry, "tran_server::ps_connector");
+}
+
+void
+tran_server::ps_connector::terminate ()
+{
+  if (m_terminate.exchange (true) == false)
+    {
+      cubthread::get_manager ()->destroy_daemon (m_daemon);
+    }
+}
+
+void
+tran_server::ps_connector::try_connect_to_all_ps (cubthread::entry &)
+{
+  /* Assume that they store PS information in the same order.
+   * TODO: combine two vectors */
+  assert (m_ts.m_connection_list.size () == m_ts.m_page_server_conn_vec.size());
+
+  for (size_t i = 0; i < m_ts.m_page_server_conn_vec.size (); i++)
```

Comment:
Nit:
```suggestion
  for (auto & connection_handler : m_ts.m_page_server_conn_vec)
```
something like that.

Reply:
We need the index to access `m_ts.m_connection_list[i]`. It will be removed anyway in the following PR for combining `m_ts.m_connection_list` and `m_ts.m_page_server_conn_vec`.